SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityScheduler.java

```diff
@@ -214,66 +217,81 @@ public void checkForDecomissions(SingularitySchedulerStateCache stateCache) {
+      final SingularityDeployKey deployKey = singularityDeployKeyListEntry.getKey();
+      final List<SingularityPendingRequest> pendingRequestsForDeploy = singularityDeployKeyListEntry.getValue();
+      final Optional<SingularityRequestWithState> maybeRequest = requestManager.getRequest(requestId);
+      final SingularityDeployStatistics deployStatistics = getDeployStatistics(deployKey.getRequestId(), deployKey.getDeployId());
 
       if (!isRequestActive(maybeRequest)) {
-        LOG.debug("Pending request {} was obsolete (request {})", pendingRequest, SingularityRequestWithState.getRequestState(maybeRequest));
+        LOG.debug("Pending request {} was obsolete (request {})", requestId, SingularityRequestWithState.getRequestState(maybeRequest));
         obsoleteRequests++;
-        requestManager.deletePendingRequest(pendingRequest);
+        for (SingularityPendingRequest pendingRequest : pendingRequestsForDeploy) {
+          requestManager.deletePendingRequest(pendingRequest);
+        }
         continue;
       }
 
-      Optional<SingularityRequestDeployState> maybeRequestDeployState = deployManager.getRequestDeployState(pendingRequest.getRequestId());
-      Optional<SingularityPendingDeploy> maybePendingDeploy = deployManager.getPendingDeploy(maybeRequest.get().getRequest().getId());
-
-      final SingularityRequest updatedRequest;
-      if (maybePendingDeploy.isPresent() && pendingRequest.getDeployId().equals(maybePendingDeploy.get().getDeployMarker().getDeployId())) {
-        updatedRequest = maybePendingDeploy.get().getUpdatedRequest().or(maybeRequest.get().getRequest());
-      } else {
-        updatedRequest = maybeRequest.get().getRequest();
-      }
-
-      if (!shouldScheduleTasks(updatedRequest, pendingRequest, maybePendingDeploy, maybeRequestDeployState)) {
-        LOG.debug("Pending request {} was obsolete (request {})", pendingRequest, SingularityRequestWithState.getRequestState(maybeRequest));
-        obsoleteRequests++;
-        requestManager.deletePendingRequest(pendingRequest);
-        continue;
-      }
+      Optional<SingularityRequestDeployState> maybeRequestDeployState = deployManager.getRequestDeployState(requestId);
+      Optional<SingularityPendingDeploy> maybePendingDeploy = deployManager.getPendingDeploy(requestId);
+      List<SingularityTaskId> matchingTaskIds = getMatchingTaskIds(stateCache, maybeRequest.get().getRequest(), deployKey);
 
-      final List<SingularityTaskId> matchingTaskIds = getMatchingTaskIds(stateCache, updatedRequest, pendingRequest);
+      List<SingularityPendingRequest> effectivePendingRequests = new ArrayList<>();
+      pendingRequestsForDeploy.sort(Comparator.comparingLong(SingularityPendingRequest::getTimestamp));
+      for (SingularityPendingRequest pendingRequest : pendingRequestsForDeploy) {
+        final SingularityRequest updatedRequest = updatedRequest(maybePendingDeploy, pendingRequest, maybeRequest.get());
```

Comment:
We should move this earlier on and make sure all methods that use the SingularityRequest object are using this one. i.e. in the old code `getMatchingTaskIds` was also called with the updated one instead of `maybeRequest.get().getRequest()`. Maybe put this up right after the isRequestActive check?